spring/spring4/spring4-common/src/main/java/org/infinispan/spring/provider/CacheDelegate.java

```diff
@@ -71,13 +77,31 @@ public ValueWrapper get(final Object key) {
 
    @Override
    public <T> T get(Object key, Callable<T> valueLoader) {
-      return (T) nativeCache.computeIfAbsent(key, keyToBeInserted -> {
+      ReentrantLock lock = null;
+      T value = (T) nativeCache.get(key);
+      if (value == null) {
+         lock = synchronousGetLocks.computeIfAbsent(key, k -> new ReentrantLock());
+         lock.lock();
          try {
-            return valueLoader.call();
-         } catch (Exception e) {
-            throw ValueRetrievalExceptionResolver.throwValueRetrievalException(key, valueLoader, e);
+            if ((value = (T) nativeCache.get(key)) == null) {
+               try {
+                  T newValue = valueLoader.call();
+                  // we can't use computeIfAbsent here since in distributed embedded scenario we would
+                  // send a lambda to other nodes. This is the behavior we want to avoid.
+                  value = (T) nativeCache.putIfAbsent(key, newValue);
+                  if (value == null) {
+                     value = newValue;
+                  }
+               } catch (Exception e) {
+                  throw ValueRetrievalExceptionResolver.throwValueRetrievalException(key, valueLoader, e);
+               }
+            }
+         } finally {
+            lock.unlock();
+            synchronousGetLocks.remove(key);
```

Comment:
I feel we should have a comment here to explain that removing the lock from `synchronousGetLocks` allows 2 threads to invoke `valueLoader` in parallel, but that's fine because the `get(key, valueLoader)` javadoc allows it.